### DISCUSSION TITLE
Support Occlum statically-linked mode

### DISCUSSION THREAD

**Discussion by kailun-qin** [occlum/occlum#295] [Ideas]
# Motivation
Comparing to Occlum LibOS mode, this statically-linked mode is targeting:
* Smaller TCB
-- for users who do not wish to include the whole LibOS
-- achieved by linking only to the used functions
* Potential better performance for specific apps/libs
-- eliminate unnecessary overhead (register switches) for some syscalls
-- LibOS grpc uds

# Drafted Design
## High Level Design
      +--------------------------+                     +---------------+
      |  Enclave (.so)           |                     | untrusted app |
      | +-------------+          |                     | (occlum-run)  |
      | | app/lib (.a)| <----------------------------- +---------------+
      | +-------------+------+   |  ecall_invoke_main  +---------------+
      | | sgx_tcxx    |      |   |  ecall_invoke_func  | occlum-pal.so |
      | | sgx_tstdc   | LibOS|   |                     +---------------+
      | | occlum musl | (.a) |   |                             ^
      | | ...         |      |   |                             |
      | +-------------+      |   |                             |
      | | occlum syscalls    | --------------------------------+
      | +--------------------+   |  ocalls/hostcalls
      +--------------------------+

- app/lib statically linked to Occlum LibOS, its deps (sgx_tcxx，sgx_tstdc，rust sgx sdk etc.) and musl libc, with link priority: sgx_tcxx/sgx_tstdc > musl libc
- exposing ecall_invoke_main/ecall_invoke_func stub functions for the untrusted
- ocalls/hostcalls on LibOS-unhandled syscalls

##  Work flow
1. pal_register_sig_handlers
2. pal_init_enclave
3. occlum_ecall_init
i. init log_level
ii. register_exception_handlers
iii. init interrupt
4. occlum_ecall_invoke_main/occlum_ecall_invoke_func
i. prepare environment
-- fake/dummy task
-- prepare syscall stack
-- save & set task (%gs)
ii. app: main
-- sgx_tcxx/sgx_tstdc
-- musl libc
-- syscall
-- Occlum syscall handler
-- do_syscall
-- (ocalls/hostcalls)
-- do_sysret
iii. restore environment
-- restore %gs
5. destroy_enclave

# PoC
## Changes Possibly Required
1. app
i. main/func as normal
ii. built to .a
2. linker
i. link order (-Wl,--start-group ...-lsgx_tcxx -lsgx_tstdc -lmusl_c... -Wl,--end-group)
3. musl
i. weak for all funcs
ii. set __occlum_entry
-- weak extern hidden size_t __occlum_syscall_linux_abi(void) {}
-- __occlum_entry = (size_t)__occlum_syscall_linux_abi
-- init enclave -> special ecall, construct function
4. occlum_ecall_invoke_main
i. fake dummy task
-- prepare syscall stack
-- set and restore %gs
ii. no need for __exec_task (save kernel fsbase, user user fsbase & stack, jmp to user code)
5. syscall
i. no need for wrfsbase
6. fd
i. used to bond w/ process
ii. skip searching from the file table
iii. use fd directly from syscalls
7. skip checkers of primitives (are within the memory of the user process)
i. from_user::check_array, from_user::check_mut_ptr etc.

## Reference Implementation
[Occlum PR](https://github.com/occlum/occlum/pull/294)
[musl PR](https://github.com/occlum/musl/pull/8)

# TODO / Open for Discussions
- syscall definitions by macros with syscall num information, which is used for dead code elimination
- Compatibility with Occlum LibOS mode
- Modules which need further adaptations, e.g. fs, process, signal etc.

**Comment by guzongmin** [comment]
We could continue to evaluate ROI of the static mode.
The most important thing is: Needs more use cases which require this mode.
